### PR TITLE
TypeScript: one request lifecycle, so every attempt is observed exactly once

### PR DESCRIPTION
**Pull request by jeremy**
Four defects in the client middleware, all rooted in the same thing: the retry loop re-issues through raw `fetch()`, which leaves the middleware chain, and nothing owned the lifecycle across that boundary.

## The defects

**1. The retry replayed the wrong body.** The body cache was keyed on `` `${method}:${url}:${Date.now()}` ``. Two concurrent mutations to the same URL land in the same millisecond routinely — a `Promise.all` batch, two workers on one record — and shared one cache slot. Whichever response arrived first deleted it, so the other request's retry replayed the wrong bytes or an empty body. That is silent data corruption on a retry, not a bookkeeping error.

**2. `onRequestEnd` never fired on a rejected initial fetch.** openapi-fetch skips `onResponse` entirely on that path and routes only to `onError`, which no middleware implemented. During a DNS outage or a stall, every request fired a start and no end: in-flight climbs, error rate reads zero — the opposite of what the hook was wired up for.

**3. Body and timing state leaked** on the retry and error paths, pruned only in `onResponse`.

**4. The raw retry got no hooks at all**, so attempt 2 was invisible; and `onRetry` reported the attempt that had just *failed* where SPEC §7 asks for the *upcoming* one. Go, Python, Ruby and Kotlin all pass `(failed, failed + 1)`; TypeScript passed `(1, 1)` — in **both** call sites, `client.ts` and the multipart loop in `services/base.ts`.

## The design

One `RequestLifecycle` keyed on openapi-fetch's per-request `id`, replacing two independent state maps. That id is minted once per logical request and threaded unchanged through `onRequest`/`onResponse`/`onError`, so it survives the auth middleware swapping the `Request` object, and it never touches the wire.

Two correlation headers go away with it. `X-SDK-Request-Id`, and `X-Request-Id` — the latter mattering because **SPEC reserves that name for the BC3 *response* header** feeding `error.request_id` (read at `services/base.ts:506`), so sending an SDK-internal value under it was a collision. Verified nothing in any SDK, spec, or conformance fixture referenced either header.

`finalize()` is idempotent, which is what makes split ownership safe:

- the retry middleware ends the failed attempt **before sleeping**, so a slow backoff can't leave an attempt open
- it announces the upcoming attempt, begins it, and finalizes its own raw fetch directly
- the hooks middleware's `onResponse` runs later for the same logical request and becomes a harmless no-op
- a failed retry fetch is caught, finalized with `statusCode: 0`, and **rethrown with its identity intact** — necessary because `onError` only covers the initial fetch
- body state is dropped in `finally`, plus in an `onError`, since a rejected initial fetch never reaches `onResponse`

`id` deliberately does *not* distinguish attempts, so per-attempt data nests under it.

**Kept the one-retry cap** (SPEC waiver 2B.1, `SPEC.md:444`/`468`/`472`). Lifting it is a real behavior change — 429/503 would start producing three attempts where fixtures assume two — and belongs in its own PR.

## Verification

All six new regressions were confirmed **red against the unfixed code first**:

| Regression | Failure before the fix |
|---|---|
| Concurrent same-URL bodies (with `Date.now()` pinned, both held at a barrier) | `['{"content":"BBB"}', …] to deeply equal ['{"content":"AAA"}', …]` |
| Initial network error → balanced start/end | `expected [] to have a length of 1` |
| Timeout → balanced start/end | `expected [] to have a length of 1` |
| Retry fetch fails → both attempts balanced | `expected [ 1 ] to deeply equal [ 1, 2 ]` |
| `onRetry` upcoming attempt | `expected 1 to be 2` |
| Repeated retried requests stay balanced | `length of 10 but got 5` |

I rewrote the body test partway through — it had keyed on `X-Retry-Attempt`, which this PR removes — and **re-proved it red** afterwards rather than assuming it still bit.

Then: **825 tests pass**; `make ts-check` clean; `check-retry-metadata-parity` still recognises `base.ts` under criterion 2; `check-idempotency-parity` 10/10; TypeScript drift gate clean.

## Two existing tests were changed, deliberately

Both asserted the defective behavior, so I want these visible rather than buried in the diff:

- `integration.test.ts` expected `onRequestStart` exactly once, with the comment *"The retry uses raw fetch, so it bypasses the hooks middleware."* It now asserts one start and one end per attempt, with `429` then `200`.
- `client.test.ts` pinned `onRetry`'s second argument to `1`. Now `2`, with `info.attempt: 1`.

## Not in scope

Swift has the same `(1, 1)` numbering and should be tracked separately — deliberately not touched here to keep this PR one language.

---
## Summary by cubic
Unifies the request lifecycle so each attempt is observed exactly once. Fixes leaks when no hooks are set, reports retried conditional GETs as cached 200s with `fromCache`, cancels failed response streams before backoff, and accounts for auth refresh failures during retries.

- **Bug Fixes**
  - Retries replay the correct body for concurrent same-URL mutations.
  - `onRequestEnd` now fires on initial network errors/timeouts (statusCode 0 with the original error).
  - Lifecycle middleware is registered even with no hooks; per-request state is always released.
  - Cached conditional GET retries that return 304 are finalized after cache transformation and reported as 200 with `fromCache`.
  - Failed 429/503 responses cancel their body stream before sleeping, so throttled clients don’t hold open connections during backoff.
  - Attempt 2 begins before the auth refresh; if the refresh throws, the attempt is finalized with statusCode 0 and released so announced retries are always accounted.

- **Refactors**
  - Introduced a shared `RequestLifecycle` keyed by `openapi-fetch` request `id`; removes `X-SDK-Request-Id` and `X-Request-Id`.
  - Retry middleware begins and announces attempts; successful retries finalize downstream so cache transforms are reflected. Errors during the retry block finalize and release state.
  - Aligned attempt semantics with the spec: `RequestInfo.attempt` = failed attempt; `onRetry` arg = upcoming. `consoleHooks` logs the upcoming attempt verbatim.
  - Renamed and updated comments: lifecycle middleware replaces the old hooks middleware; comments now reflect the lifecycle and order.

<sup>Written for commit 194faf32ce8f34617a35e48bbae0ae5f8d96916f. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/basecamp/basecamp-sdk/pull/503?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: typescript/src/client.ts [hunk 1]
```diff
@@ -492,78 +501,155 @@ function createAuthMiddleware(authStrategy: AuthStrategy, userAgent: string, req
...
+    url: string,
+    failedAttempt: number,
+    error: Error,
+    delayMs: number,
+  ): void {
+    this.emit(() =>
+      this.hooks?.onRetry?.(
+        { method, url, attempt: failedAttempt },
+        failedAttempt + 1,
+        error,
```
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/src/client.ts:599]: Fixed in `e3b2c6db` — correct, and a good catch on the coupling. `hooks.ts:290` printed `attempt + 1`, which read correctly only because the callers were passing the attempt that had just failed. Now that both call sites pass the upcoming attempt, it said "attempt 3" while attempt 2 was in flight. It prints the supplied value verbatim.

Its test was pinning the compensating pair rather than either rule — it already passed `2` and asserted `3` — so that is updated too.

File: typescript/src/client.ts [hunk 2]
```diff
@@ -889,123 +975,120 @@ function getRetryConfigForRequest(method: string, url: string): RetryConfig {
...
+        });
+
+        // Refresh auth header for retry (token may have been refreshed since initial request)
+        if (authStrategy) {
+          await authStrategy.authenticate(retryRequest.headers);
+        }
+
+        // This raw fetch bypasses the middleware chain, so nothing downstream will
+        // observe it — this middleware owns the whole lifecycle of the retry.
+        lifecycle.begin(id, method, url, failedAttempt + 1);
```
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/src/client.ts]: Fixed in `e3b2c6db` — real, and a leak I introduced while fixing a leak.

Root cause was the registration, not the retry path: `createLifecycleMiddleware` (was `createHooksMiddleware`) is the only thing that releases on the success path, and it was registered only when `hooks` was truthy — while the retry middleware records an attempt regardless of whether anyone is listening. So it is now registered unconditionally. Every emit is optional-chained, so with no hooks it observes nothing and exists purely to release.

That also closes the wider version you both point at: the `catch` now wraps the whole retry block rather than just the fetch, so a throw from `authStrategy.authenticate` or from `sleep` finalizes and releases too.

One thing I want to be straight about: the added no-hooks test exercises the retry path end to end but does **not** assert the release itself. That state is private and not reachable through the public API, so I did not want to fake an assertion or widen the export surface for a test. The fix there is structural.

File: typescript/src/client.ts [hunk 3]
```diff
@@ -889,123 +975,120 @@ function getRetryConfigForRequest(method: string, url: string): RetryConfig {
...
+          lifecycle.finalize(id, method, url, {
+            statusCode: 0,
+            error: error instanceof Error ? error : new Error(String(error)),
+          });
+          lifecycle.release(id);
+          throw error;
+        }
+      } finally {
+        bodyCache.delete(id);
       }
```
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/src/client.ts]: Fixed in `e3b2c6db` — real, and a leak I introduced while fixing a leak.

Root cause was the registration, not the retry path: `createLifecycleMiddleware` (was `createHooksMiddleware`) is the only thing that releases on the success path, and it was registered only when `hooks` was truthy — while the retry middleware records an attempt regardless of whether anyone is listening. So it is now registered unconditionally. Every emit is optional-chained, so with no hooks it observes nothing and exists purely to release.

That also closes the wider version you both point at: the `catch` now wraps the whole retry block rather than just the fetch, so a throw from `authStrategy.authenticate` or from `sleep` finalizes and releases too.

One thing I want to be straight about: the added no-hooks test exercises the retry path end to end but does **not** assert the release itself. That state is private and not reachable through the public API, so I did not want to fake an assertion or widen the export surface for a test. The fix there is structural.

File: typescript/src/client.ts [hunk 4]
```diff
@@ -889,123 +975,120 @@ function getRetryConfigForRequest(method: string, url: string): RetryConfig {
...
+          await authStrategy.authenticate(retryRequest.headers);
+        }
+
+        // This raw fetch bypasses the middleware chain, so nothing downstream will
+        // observe it — this middleware owns the whole lifecycle of the retry.
+        lifecycle.begin(id, method, url, failedAttempt + 1);
+        try {
+          const retryResponse = await fetch(retryRequest);
+          lifecycle.finalize(id, method, url, { statusCode: retryResponse.status });
+          return retryResponse;
```
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/src/client.ts]: Fixed in `e3b2c6db` — you traced this exactly, including why idempotence made it unrecoverable.

The retry no longer finalizes a successful attempt. It begins attempt 2 and returns the response, leaving finalization to the downstream pass so the cache middleware can rewrite a 304 into a cached 200 first. The error path still finalizes locally, since `onError` covers only the initial fetch.

Added a regression for your exact scenario — cached conditional GET, 429, retry returns 304 — and confirmed it red against the previous commit: `expected 304 to be 200`.

**Comment by jeremy** [comment]
Zero inline threads outstanding, but Copilot's last pass carried **two suppressed low-confidence comments** and both were correct — so I've acted on them rather than letting the empty thread count stand in for a clean review.

They flagged comments left stale by renaming `createHooksMiddleware` → `createLifecycleMiddleware` and making it unconditional. Nine references in total, including a genuine artifact I'd missed: an orphaned

```
/** Tracks request timing for hooks */
```

stranded directly above its replacement doc comment, documenting a `RequestTiming` map this PR deleted. Also retitled the section from "Hooks Middleware" to "Request Lifecycle" and corrected the registration comment, which still listed "then hooks" in the middleware order and described the reverse-order response pass as ending at the hooks middleware.

Comment-only — I verified the diff contains no non-comment lines. 827 tests still pass, `ts-check` clean. Pushed as `e1af941a`.

This is the second time in this PR that a *suppressed* Copilot comment turned out to be right, so it seems worth saying plainly: low confidence is not the same as low value, and an empty unresolved-thread count says nothing about the suppressed set.

File: typescript/src/client.ts [hunk 5]
```diff
@@ -889,123 +977,124 @@ function getRetryConfigForRequest(method: string, url: string): RetryConfig {
...
 
-      // Get cached body for methods that may have one
-      let body: ArrayBuffer | null = null;
-      if (requestId && bodyCache.has(requestId)) {
-        const cachedBody = bodyCache.get(requestId);
-        if (cachedBody) {
-          body = cachedBody;
-        }
-      }
+        await sleep(delay);
```
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/src/client.ts:1056]: Fixed in `55272e79` — real, and the inconsistency is the tell: `services/base.ts:205` already cancels at exactly this point, with the comment "Drain response body before retry to free resources and enable connection reuse". The middleware was the odd one out.

Cancels before sleeping, ignoring errors since the body may already be consumed or closed. Optional chaining short-circuits the whole chain when there is no body, and the promise is voided so a rejection cannot surface unhandled.

Also added a retry test whose failed response actually carries a body — every other retry test in that file uses a null-body response, so none of them reached this branch. It pins that cancelling cannot break the retry; connection reuse itself is not observable from the harness.

**Comment by jeremy** [comment]
Two more rounds addressed. Both came from bots and both were real.

**Codex — drain the discarded response** (`55272e79`). A retryable 429/503 was dropped without cancelling its body, so a throttled client held one open stream and connection per in-flight retry across the whole backoff. The tell was the inconsistency: `services/base.ts:205` already cancels at exactly this point, with a comment about connection reuse. Added a retry test whose failed response actually carries a body — every other retry test in the file uses a null-body response, so none of them reached that branch.

**Copilot, suppressed — attempt 2 could be announced and then vanish** (`61eded01`). `lifecycle.begin` ran *after* the auth refresh, so a throw from `authStrategy.authenticate` left the `catch` with nothing live to finalize: state was still the finalized attempt 1, so `finalize` no-oped. Hooks saw `onRetry(upcoming=2)` and then no start and no end for it. That is the same blind spot this PR exists to close, one step later in the sequence.

`begin` now runs just after the backoff — after the sleep so the attempt's duration measures the request rather than the wait, before the request build and auth refresh so anything throwing lands on a live attempt. The regression injects an auth strategy that succeeds initially and throws on refresh; confirmed red against the previous commit with `expected [ 1 ] to deeply equal [ 1, 2 ]`.

**829 tests pass**, `ts-check` and retry-metadata parity clean.

Worth stating, because it has now happened three times on this PR: every one of the last three findings arrived as a **suppressed low-confidence** Copilot comment or a P2 Codex note, while the inline unresolved-thread count sat at zero. The stale-comment cleanup, and now a genuine lifecycle hole, were both in that suppressed set. A zero thread count is not a clean review here.

**Comment by jeremy** [comment]
Fixed in `194faf32` — and you were right that it was no regression at all. The test passed unchanged with `response.body?.cancel()` deleted, so it only ever proved that a retry with a body still succeeds.

My error was in the reasoning, not the code: I concluded the behaviour was unobservable because *connection reuse* is, and stopped there. The cancellation **call** is plainly observable. Now spies on `ReadableStream.prototype.cancel` and asserts it happened, renamed to "cancels the failed response's stream before retrying" to say what it checks.

Confirmed load-bearing by deleting only the cancel line: `expected "cancel" to be called at least once`. No unrelated stream cancels polluted the spy, so a plain `toHaveBeenCalled()` is tight enough — worth stating, since if MSW or undici had cancelled other streams the assertion would have been satisfied for the wrong reason.

829 tests pass; `ts-check` clean. Head `194faf32`.

On the audit rule — agreed, and I've recorded it as a standing practice rather than a note on this PR. Suppressed comments are review findings stored outside the thread ledger: they never become threads, so they never appear in `reviewThreads` and never move the unresolved count. Three of this PR's real defects arrived that way at 0 unresolved. The sweep is now both surfaces — inline comments via `reviewThreads`, and every review body grepped for the suppressed-comments block, per reviewer, on the current head.
